ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util.hs

```diff
@@ -374,3 +388,48 @@ fib n = round $ phi ** fromIntegral n / sq5
 eitherToMaybe :: Either a b -> Maybe b
 eitherToMaybe (Left _)  = Nothing
 eitherToMaybe (Right x) = Just x
+
+{-------------------------------------------------------------------------------
+  Fuse code, i.e. regions of code that will throw an exception if accessed
+  concurrently.
+-------------------------------------------------------------------------------}
+
+-- | A WithIntactFuse expression is one of which a single instance can be
+-- evaluated at any given moment in the execution of the program.
+newtype WithIntactFuse m a = WithIntactFuse (m a)
+  deriving newtype (Functor, Applicative, Monad, MonadThrow, MonadCatch)
+
+instance MonadTrans WithIntactFuse where
+  lift = WithIntactFuse
+
+-- | Declare that an action is not safe to be run concurrently, i.e. two
+-- concurrent accesses should throw an exception.
+blowFuse = WithIntactFuse
+blowFuse :: m a -> WithIntactFuse m a
+
+-- | Lock required to run a 'WithIntactFuse'
+data Fuse m = Fuse String (MVar m ())
+
+newFuse :: MonadMVar m => String -> m (Fuse m)
+newFuse name = Fuse name <$> newMVar ()
+
+-- | Run the 'WithIntactFuse' action throwing an exception if it is already
+-- running in a separate thread.
+blowAndResetFuse ::
+     (MonadThrow m, MonadMVar m)
+  => Fuse m
+  -> WithIntactFuse m a
+  -> m a
+blowAndResetFuse (Fuse name m) (WithIntactFuse io) = do
+    () <- onlyOne =<< tryTakeMVar m
+    a <- io
+    onlyOne . bool Nothing (Just ()) =<< tryPutMVar m ()
+    pure a
+  where
+    onlyOne = maybe (throwIO $ FuseCodeAccessedConcurrently name) pure
```

Comment:
As this is currently written without `bracket`, `tryPutMVar` does not happen when `io` throws. In fact, *any* subsequent `blowAndResetFuse` with the same `Fuse` will then later fail. I think this is fine for our use case, but it might be at least worth to document this behavior as it is a bit subtle.

Basically: A `Fuse` (as of the state of the PR right now) is not only about making sure that a piece of code is not run concurrently, but also about making sure that it will always fail if it failed once.

Reply:
Good point!

Reply:
Done